src/openfl/text/TextField.hx

```diff
@@ -3013,9 +2977,47 @@ class TextField extends InteractiveObject implements IShaderDrawable {
 			
 		}
 		
-		replaceSelectedText (value);
+		__inputText (value);
+		
+	}
+	
+	
+	private function __inputText (value:String):Void {
+		
+		var event = new TextEvent (TextEvent.TEXT_INPUT, true, true, value);
+		dispatchEvent (event);
+		
+		if (!event.isDefaultPrevented ()) {
+			
+			replaceSelectedText (value);
+			dispatchEvent (new Event (Event.CHANGE, true));
+			
+		}
 		
-		dispatchEvent (new Event (Event.CHANGE, true));
+	}
+	
+	
+	private function window_onTextCopy (provideData:CopyDataProvider):Void {
+		
+		if (__caretIndex != __selectionIndex) {
+			
+			provideData (__text.substring (__caretIndex, __selectionIndex));
```

Comment:
This goes in the direction of my lime comment to remove the callback and just call `Clipboard.setData (__text.substring (__caretIndex, __selectionIndex), false)` here.

Reply:
See my reply there and let's keep the discussion there :D